library/std/src/sys/fs/windows.rs

```diff
@@ -848,6 +904,366 @@ impl File {
+    f(path)
+}
+
+fn run_path_with_utf16<T, P: AsRef<Path>>(
+    path: P,
+    f: &dyn Fn(&[u16]) -> io::Result<T>,
+) -> io::Result<T> {
+    let utf16: Vec<u16> = path.as_ref().as_os_str().encode_wide().collect();
+    f(&utf16)
+}
+
+impl Dir {
+    pub fn new<P: AsRef<Path>>(path: P) -> io::Result<Self> {
+        let mut opts = OpenOptions::new();
+        opts.read(true);
+        run_path_with_wcstr(path.as_ref(), &|path| Self::new_with_native(path, &opts))
+    }
+
+    pub fn new_with<P: AsRef<Path>>(path: P, opts: &OpenOptions) -> io::Result<Self> {
+        run_path_with_wcstr(path.as_ref(), &|path| Self::new_with_native(path, &opts))
+    }
+
+    pub fn new_for_traversal<P: AsRef<Path>>(path: P) -> io::Result<Self> {
+        run_path_with_wcstr(path.as_ref(), &|path| Self::new_native(path))
+    }
+
+    pub fn open<P: AsRef<Path>>(&self, path: P) -> io::Result<File> {
+        let mut opts = OpenOptions::new();
+        opts.read(true);
+        let path = path.as_ref();
+        if path.is_absolute() {
+            return File::open(path, &opts);
+        }
+        let path = path.as_os_str().encode_wide().collect::<Vec<_>>();
+        self.open_native(&path, &opts, false).map(|handle| File { handle })
+    }
+
+    pub fn open_with<P: AsRef<Path>>(&self, path: P, opts: &OpenOptions) -> io::Result<File> {
+        let path = path.as_ref().as_os_str().encode_wide().collect::<Vec<_>>();
+        self.open_native(&path, &opts, false).map(|handle| File { handle })
+    }
+
+    pub fn open_dir<P: AsRef<Path>>(&self, path: P) -> io::Result<Self> {
+        let mut opts = OpenOptions::new();
+        let path = path.as_ref().as_os_str().encode_wide().collect::<Vec<_>>();
+        opts.read(true);
+        self.open_native(&path, &opts, true).map(|handle| Self { handle })
+    }
+
+    pub fn open_dir_with<P: AsRef<Path>>(&self, path: P, opts: &OpenOptions) -> io::Result<Self> {
+        let path = path.as_ref().as_os_str().encode_wide().collect::<Vec<_>>();
+        self.open_native(&path, &opts, true).map(|handle| Self { handle })
+    }
+
+    pub fn create_dir<P: AsRef<Path>>(&self, path: P) -> io::Result<()> {
+        let mut opts = OpenOptions::new();
+        opts.write(true);
+        opts.create_new(true);
+        run_path_with_utf16(path, &|path| self.create_dir_native(path, &opts).map(|_| ()))
+    }
+
+    pub fn remove_file<P: AsRef<Path>>(&self, path: P) -> io::Result<()> {
+        run_path_with_utf16(path, &|path| self.remove_native(path, false))
+    }
+
+    pub fn remove_dir<P: AsRef<Path>>(&self, path: P) -> io::Result<()> {
+        run_path_with_utf16(path, &|path| self.remove_native(path, true))
+    }
+
+    pub fn rename<P: AsRef<Path>, Q: AsRef<Path>>(
+        &self,
+        from: P,
+        to_dir: &Self,
+        to: Q,
+    ) -> io::Result<()> {
+        run_path_with_wcstr(to.as_ref(), &|to| self.rename_native(from.as_ref(), to_dir, to, false))
```

Comment:
This will produce an absolute path rather than a relative one. You'll want to use `run_path_with_utf16` like the other functions.